src/util/dsp/levinson.cpp

```diff
@@ -0,0 +1,103 @@
+//**************************************************************************
+//
+//	$file: main.c $
+//	$author: Martin Fouilleul $
+//	$date: 28/11/2018 $
+//	$description: a simple Toeplitz system solver$
+//
+//**************************************************************************
+
+#include<string.h> // memset()
+
+namespace otto::dsp
+{
+//--------------------------------------------------------------------------
+//
+//	SolveToeplitz() is a (somewhat naive) implementation of the Levinson-Durbin algorithm for solving the Topelitz system Mx = y
+//
+//	arguments :
+//		- size : the size of the Toeplitz matrix and of vectors x and y
+//		- toeplitz : a buffer containing the first column of the matrix from bottom to top, then the first line from left to right.
+//			     eg. if the matrix is :
+//
+//			     [T0 T1   ...  Tn-1]
+//			     [T-1 T0  ...  Tn-2]
+//			     [...              ]
+//			     [T-n+1   ...    T0]
+//
+//			     you should pass the vector [T-n+1 T-n+2 ... T0 T1 ... Tn-1] of length (size+size-1)
+//
+//		- y : the vector at the right-hand side of the equation
+//		- x : result vector (that is, the x unknown at the left-hand side of the equation)
+//		- scratchBuffer : a block of memory used to hold intermediate computations. You must pass a buffer of size at least 4*size*sizeof(float).
+//
+//--------------------------------------------------------------------------
+
+void SolveToeplitz(unsigned int size, const float* toeplitz, const float* y, float* x, char* scratchBuffer)
+{
+	//NOTE(martin): Clear and organize the scratch memory
+	memset(x, 0, size*sizeof(float));
+	memset(scratchBuffer, 0, 4*size*sizeof(float));
+
+	float* forwardPrev  = (float*)scratchBuffer; scratchBuffer += size*sizeof(float);
+	float* backwardPrev = (float*)scratchBuffer; scratchBuffer += size*sizeof(float);
+	float* forward  = (float*)scratchBuffer; scratchBuffer += size*sizeof(float);
+	float* backward = (float*)scratchBuffer; scratchBuffer += size*sizeof(float);
```

Comment:
Use this instead ?
```cpp
std::vector<float> forwardPrev(size, 0.0);
std::vector<float> backwardPrev(size, 0.0);
std::vector<float> forward(size, 0.0);
std::vector<float> backward (size, 0.0);
```

Reply:
Again, it wouldn't prevent size being different from toeplitz' length, so the bound checking provided by std::vector are mostly useless in this case. 
Furthermore, this costs four allocations and four deallocations each time the function is called, although we only need a scratch buffer to do some computation. With the lower level version we can allocated once for a given size and always pass the same buffer.
So I understand the rationale behind adapting the code to fit the style of the rest of the codebase, but I just want to point out that in this specific case it comes with a performance cost.
Maybe passing scratchBuffer as a gsl::span (or even as a vector reference) could be a better solution if you want to do the bound checking explicitly ?

Reply:
Then maybe allocate the scratchpad vector in the LPC engine once, and pass it as a span in the function signature ? I can see the rationale.

I imagine the Engine would end up looking kinda like that:
```cpp
namespace otto::engines {
struct LPC : EffectEngine {
//...
private:
    std::vector<float> scratchpadBuffer;
//...
public:
    audio::ProcessData<2> Wormhole::process(audio::ProcessData<2> data)
    {
        //...
        auto toeplitz = autocovariance(mono_data, size);
        std::vector<float> y(size, 0.0);
        y[0] = 1;
        auto v = solveToeplitz(toeplitz, y);
        float sigma_2 = 1/v[0];
        auto estimated_coefficients = -sigma_2 * v;
        // Now make the filter with these coefficients
        // and filter a source (noise or pulse train)
    }
```

Reply:
I pushed another commit with the simplifications discussed above, but regarding your suggestions, it seems like I can't compile GSL and most of the latest cpp fanciness :/
I don't feel like messing around with my toolchain right now, so maybe you could pull it and apply your suggested modifications ?

Reply:
@nebularnoise I added you as a contributor to my fork of the repo, so now you can make edits to the PR directly if you  want.